src/hotspot/share/gc/g1/heapRegion.inline.hpp

```diff
@@ -181,8 +181,15 @@ inline size_t HeapRegion::block_size(const HeapWord* p, HeapWord* const pb) cons
 }
 
 inline void HeapRegion::reset_compacted_after_full_gc(HeapWord* new_top) {
-  assert(!is_pinned(), "must be");
+  set_top(new_top);
+  // After a compaction the mark bitmap in a non-pinned regions is invalid.
+  // But all objects are live, we get this by setting TAMS to bottom.
+  init_top_at_mark_start();
+
+  reset_after_full_gc_common();
+}
 
+inline void HeapRegion::reset_compacted_humongous_after_full_gc(HeapWord* new_top) {
```

Comment:
This method could probably just call `reset_compacted_after_full_gc` as they are identical. (I am not suggestion to remove this method, although I'm not completely sure it's useful).